Add documentation for Kueue visibility on-demand (#1381)
kubernetes-sigs/kueue: site/content/en/docs/tasks/monitor_pending_workloads/pending_workloads_on_demand.md

---
title: "Pending Workloads on-demand"
date: 2023-12-04
weight: 3
description: >

> #   Monitor pending workloads.

Reviewer: ditto.

From version v0.6.0, Kueue provides the ability for a batch administrators to monitor
the pipeline of pending jobs, and help users to estimate when their jobs will
start.

## Before you begin

Make sure the following conditions are met:

- A Kubernetes cluster is running.
- The kubectl command-line tool has communication with your cluster.
- [Kueue is installed](/docs/installation) in version v0.6.0 or later.

## Enabling feature VisibilityOnDemand

VisibilityOnDemand is an `Alpha` feature disabled by default, check the [Change the feature gates configuration](/docs/installation/#change-the-feature-gates-configuration) section of the [Installation](/docs/installation/) for details.

## Monitor pending workloads on demand

> _Available in Kueue v0.6.0 and later_

To install a simple setup of cluster queue, run the following command:

```shell
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/admin/single-clusterqueue-setup.yaml
```

To install two local queues, run the following command:

```shell
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/admin/localqueues-setup.yaml

> ```

Reviewer: I believe that we should provide the only minimum case. So setting a localqueue by `single-clusterqueue-setup.yaml` should be enough.

Or is there any reason that we should have multiple localqueues?

Reviewer 2: Yeah,  I don't have a strong view, maybe we can go with minimal setup in the first iteration.

One reason to have 2 localqueues is to demonstrate that the positions in ClusterQueue might be interleaving between two LocalQueues, for this purpose we could just install another local queue from a dedicated file.

Author: My intention was to demonstrate two facts:
1) The position of Workloads across LocalQueues is consistent, 
2) A user requesting Workloads belonging to one LocalQueue does not have any information about pending Workloads in other LocalQueues.

Reviewer 2: Sounds reasonable to me, WDYT @tenzen-y ?

Now, let's create a job that will block cluster queue, so no job is admitted

```shell
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/jobs/blocking-job.yaml
```

And a few job beloning to LocalQueue `team-a-queue`

```shell
for i in {1..5}; do kubectl create -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/jobs/team-a-job.yaml; done
```

and to LocalQueue `team-b-queue`

```shell
for i in {1..5}; do kubectl create -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/jobs/team-b-job.yaml; done
```

### Cluster Queue visibility

To view pending workloads in ClusterQueue run the following command:

```shell
kubectl get --raw /apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/pendingworkloads
```

You can pass optional query parameters:

> - limit `<int>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
> - offset `<int>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.

Reviewer: ```suggestion
- limit `<integer>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
- offset `<integer>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.
```

To view only the first 3 pending workloads use:
```shell

> kubectl get --raw /apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/pendingworkloads\?limit\=3

Reviewer: ... \?limit\=3

Why do we use the escape character "\\\" here?

Author: If one doesn't use quotation marks in the command "\" are necessary. I agree that using quotation marks and omitting "\" is a cleaner way so I'll fix that

```

To view pending workloads having at least position 3 in Cluster Queue use:
```shell
kubectl get --raw /apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/pendingworkloads\?offset\=3
```

### Local Queue visibility

Similarly to Cluster Queue, to view pending workloads in LocalQueue `team-a-queue` run the following command:

```shell
kubectl get --raw /apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/default/localqueues/team-a-queue/pendingworkloads
```

Similarly to Cluster Queue, you can pass optional query parameters:

> - limit `<int>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
> - offset `<int>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.

Reviewer: ```suggestion
- limit `<integer>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
- offset `<integer>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.
```

---
title: "Pending Workloads on-demand"
date: 2023-12-05
weight: 3
description: >
  Pending Workloads on-demand
---

This page shows you how to monitor pending workloads with VisibilityOnDemand feature.

The intended audience for this page are [batch administrators](/docs/tasks#batch-administrator), and [batch users](/docs/tasks#batch-user) for [Local Queue Visibiility section](#local-queue-visibility).

From version v0.6.0, Kueue provides the ability for a batch administrators to monitor
the pipeline of pending jobs, and help users to estimate when their jobs will
start.

## Before you begin

Make sure the following conditions are met:

- A Kubernetes cluster is running.
- The kubectl command-line tool has communication with your cluster.
- [Kueue is installed](/docs/installation) in version v0.6.0 or later.

### Enabling feature VisibilityOnDemand

VisibilityOnDemand is an `Alpha` feature disabled by default, check the [Change the feature gates configuration](/docs/installation/#change-the-feature-gates-configuration) section of the [Installation](/docs/installation/) for details.

## Monitor pending workloads on demand

> _Available in Kueue v0.6.0 and later_

> To install a simple setup of cluster queue

Reviewer 2: ```suggestion
To install a simple setup of ClusterQueue
```
ni: for consistency

        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-jg9dw",
            "uid": "fd5d1796-f61d-402f-a4c8-cbda646e2676"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 1,
      "positionInLocalQueue": 1
    },
    {
      "metadata": {
        "name": "job-sample-job-t9b8m-4e770",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-t9b8m",
            "uid": "64c26c73-6334-4d13-a1a8-38d99196baa5"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 2,
      "positionInLocalQueue": 2
    }
  ]
}
```

You can pass optional query parameters:
- limit `<integer>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
- offset `<integer>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.

> To view only 1 pending workloads use, starting from position 1 in Cluster Queue run:

Reviewer 2: ```suggestion
To view only 1 pending workloads use, starting from position 1 in ClusterQueue run:
```
nit, for consistency

        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-jg9dw",
            "uid": "fd5d1796-f61d-402f-a4c8-cbda646e2676"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 1,
      "positionInLocalQueue": 1
    },
    {
      "metadata": {
        "name": "job-sample-job-t9b8m-4e770",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-t9b8m",
            "uid": "64c26c73-6334-4d13-a1a8-38d99196baa5"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 2,
      "positionInLocalQueue": 2
    }
  ]
}
```

You can pass optional query parameters:
- limit `<integer>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
- offset `<integer>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.

> To view only 1 pending workloads use, starting from position 1 in Local Queue run:

Reviewer 2: ```suggestion
To view only 1 pending workloads use, starting from position 1 in LocalQueue run:
```
nit, for consistency

the pipeline of pending jobs, and help users to estimate when their jobs will
start.

## Before you begin

Make sure the following conditions are met:

- A Kubernetes cluster is running.
- The kubectl command-line tool has communication with your cluster.
- [Kueue is installed](/docs/installation) in version v0.6.0 or later.
- You have sufficient permissions according to [Setup RBAC section](#setup-rbac)

### Enabling feature VisibilityOnDemand

VisibilityOnDemand is an `Alpha` feature disabled by default, check the [Change the feature gates configuration](/docs/installation/#change-the-feature-gates-configuration) section of the [Installation](/docs/installation/) for details.

## Monitor pending workloads on demand

> _Available in Kueue v0.6.0 and later_

To install a simple setup of cluster queue

{{< include "examples/admin/single-clusterqueue-setup.yaml" "yaml" >}}

run the following command:

```shell
kubectl apply -f https://kueue.sigs.k8s.io/examples/admin/single-clusterqueue-setup.yaml
```

Now, let's create 10 jobs

{{< include "examples/jobs/sample-job.yaml" "yaml" >}}

using a command: 

```shell

> for i in {1..10}; do kubectl create -f https://kueue.sigs.k8s.io/examples/jobs/sample-job.yaml; done

Reviewer: ```suggestion
for i in {1..5}; do kubectl create -f https://kueue.sigs.k8s.io/examples/jobs/sample-job.yaml; done
```

I think that only 5 jobs should be enough since the below response api-server is a bit huge.

Reviewer: Or, we may be able to omit the part of the below responses.

From version v0.6.0, Kueue provides the ability for a batch administrators to monitor
the pipeline of pending jobs, and help users to estimate when their jobs will
start.

## Before you begin

Make sure the following conditions are met:

- A Kubernetes cluster is running.
- The kubectl command-line tool has communication with your cluster.
- [Kueue is installed](/docs/installation) in version v0.6.0 or later.
- You have sufficient permissions according to [Setup RBAC section](#setup-rbac)

### Enabling feature VisibilityOnDemand

VisibilityOnDemand is an `Alpha` feature disabled by default, check the [Change the feature gates configuration](/docs/installation/#change-the-feature-gates-configuration) section of the [Installation](/docs/installation/) for details.

## Monitor pending workloads on demand

> _Available in Kueue v0.6.0 and later_

To install a simple setup of cluster queue

{{< include "examples/admin/single-clusterqueue-setup.yaml" "yaml" >}}

run the following command:

```shell

> kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/kueue/main/site/static/examples/admin/single-clusterqueue-setup.yaml

Reviewer 3: would an URL like this work?
```
kubectl apply --server-side -k "github.com/kubernetes-sigs/kueue/config/default?ref=main"
```
(from our installation guide).

Author: IIUC this URL installs Kueue on a cluster, and does not create a ClusterQueue so I don't think they are equivalent

Reviewer 3: What I mean is the overall shape of the URL (it doesn't have raw). Another alternative would be to point to the website itself, similar to this `https://kueue.sigs.k8s.io/examples/jobs/sample-job.yaml`.

Author: Oh yes, these URLs work, I've changed it

        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-44nts",
            "uid": "07275e7b-f5f6-4783-8939-3ba891799222"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 5,
      "positionInLocalQueue": 5
    },
    {
      "metadata": {
        "name": "job-sample-job-f8h9t-a564c",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:04Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-f8h9t",
            "uid": "0916dca3-309e-4381-9791-b9a450d7f344"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 6,
      "positionInLocalQueue": 6
    }
  ]
}
```

You can pass optional query parameters:
- limit `<integer>` - 1000 on default. It indicates max number of pending workloads that should be fetched.
- offset `<integer>` - 0 by default. It indicates position of the first pending workload that should be fetched, starting from 0.

> To view only the first 2 pending workloads use:

Reviewer 3: ```suggestion
For example, to view only the first 2 pending workloads you can run the following command:
```

  "apiVersion": "visibility.kueue.x-k8s.io/v1alpha1",
  "metadata": {
    "creationTimestamp": null
  },
  "items": [
    {
      "metadata": {
        "name": "job-sample-job-5kb24-1a02c",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-5kb24",
            "uid": "f983df07-fdab-4e8b-81a6-a537d8225062"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 0,
      "positionInLocalQueue": 0
    },
    {
      "metadata": {
        "name": "job-sample-job-jrjfr-8d56e",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-jrjfr",
            "uid": "5863cf0e-b0e7-43bf-a445-f41fa1abedfa"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 1,
      "positionInLocalQueue": 1
    }
  ]
}
```

> To view pending workloads having at least position 5 in Cluster Queue use:

Reviewer 3: Too many examples that don't add much value. Mix this within the previous example.

  "apiVersion": "visibility.kueue.x-k8s.io/v1alpha1",
  "metadata": {
    "creationTimestamp": null
  },
  "items": [
    {
      "metadata": {
        "name": "job-sample-job-5kb24-1a02c",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-5kb24",
            "uid": "f983df07-fdab-4e8b-81a6-a537d8225062"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 0,
      "positionInLocalQueue": 0
    },
    {
      "metadata": {
        "name": "job-sample-job-jrjfr-8d56e",
        "namespace": "default",
        "creationTimestamp": "2023-12-05T15:42:03Z",
        "ownerReferences": [
          {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "name": "sample-job-jrjfr",
            "uid": "5863cf0e-b0e7-43bf-a445-f41fa1abedfa"
          }
        ]
      },
      "priority": 0,
      "localQueueName": "user-queue",
      "positionInClusterQueue": 1,
      "positionInLocalQueue": 1
    }
  ]
}
```

> To view pending workloads having at least position 5 in Local Queue use:

Reviewer 3: Same suggestion here.
